### PR TITLE
Add DV table plan transformer trait to prune the deleted rows from scan output

### PR DESCRIPTION
**Pull request by vkorukanti**
This PR is part of the feature: Support reading Delta tables with deletion vectors (more details at https://github.com/delta-io/delta/issues/1485)

Add a trait (used by `PrepareDeltaScan` to modify its output) to modify DV enabled tables to prune the deleted rows from scan output

Planner trait to inject a Filter just after the Delta Parquet scan. This transformer modifies the plan:<br>. <ul><li>Before rule: `<Parent Node> -> Delta Scan (key, value)` . Here we are reading `key`, `value` columns from the Delta table</li><li>After rule: `<Parent Node> -> Project(key, value) -> Filter (udf(__skip_row == 0) -> Delta Scan (key, value, __skip_row)`</li><li><ul><li>Here we insert a new column in Delta scan `__skip_row`. This value is populated by the Parquet reader using the DV corresponding to the Parquet file read (refer [to the change](https://github.com/delta-io/delta/pull/1542)) and it contains `0` if we want to keep the row. </li><li>The scan created also disables Parquet file splitting and filter pushdowns, because in order to generate the `__skip_row` we need to read the rows in a file consecutively in order to generate the row index. This is a drawback we need to pay until we upgrade to latest Apache Spark which contains Parquet reader changes that automatically generate the row_index irrespective of the file splitting and filter pushdowns.</li><li>The scan created also contains a broadcast variable of Parquet File -> DV File map. The Parquet reader created uses this map to find the DV file corresponding to the Parquet file.</li><li>Filter created just filters out rows with `__skip_row` equals to 0</li><li>And at the end we have a `Project` to keep the plan node output same as before the rule is applied</li></ul></li></ul>

In addition
* it adds the `deletionVector` to DeltaLog protocol objects (`AddFile`, `RemoveFile`)
* It also updates the `OptimizeMetadataOnlyDeltaQuery` to take into consideration of the DVs when calculating the row count.
* end-to-end integration of reading Delta tables with DVs in `DeletionVectorsSuite`

In following up PRs, will be adding extensive tests.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/sql/delta/PreprocessTableWithDVs.scala [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+
+    // If the table has no DVs enabled, no change needed
+    if (!deletionVectorsReadable(index.protocol, index.metadata)) return None
+
+    // See if the relation is already modified to include DV reads as part of
+    // a previous invocation of this rule on this table
+    if (fileFormat.hasDeletionVectorMap()) return None
+
+    // See if any files actually have a DV
+    val spark = SparkSession.getActiveSession.get
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/sql/delta/PreprocessTableWithDVs.scala:101]: nit: pass a SparkSession in if possible

File: core/src/main/scala/org/apache/spark/sql/delta/stats/PrepareDeltaScan.scala [hunk 2]
```diff
@@ -233,6 +215,7 @@ trait PrepareDeltaScanBase extends Rule[LogicalPlan]
       // It will fall back to just partition pruning at planning time.
       plan
     }
+    preprocessTablesWithDVs(updatedPlan)
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/sql/delta/stats/PrepareDeltaScan.scala:218]: QQ: How does this handle the `V2WriteCommand` case?

File: core/src/main/scala/org/apache/spark/sql/delta/PreprocessTableWithDVs.scala (hunk 1 above)
Comment (vkorukanti) [reply core/src/main/scala/org/apache/spark/sql/delta/PreprocessTableWithDVs.scala:101]: This is not easy to pass without major refactoring, given the whole convert logic is part of the extractor. Let me know if there is a way.

File: core/src/main/scala/org/apache/spark/sql/delta/stats/PrepareDeltaScan.scala (hunk 2 above)
Comment (vkorukanti) [reply core/src/main/scala/org/apache/spark/sql/delta/stats/PrepareDeltaScan.scala:218]: I am assuming you are talking about the case when stats are disabled?

I think there are existing issues around when stats are disabled. Planning to look at it once this DV project is complete. Currently when stats are disabled, we create `TahoeLogFileIndex` which doesn't have a pinned snapshot. Also the subquery handling part is missing.

File: core/src/main/scala/org/apache/spark/sql/delta/SubqueryTransformerHelper.scala [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+ * not be transformed.
+ */
+trait SubqueryTransformerHelper {
+
+  /**
+   * Transform all nodes matched by the rule in the query plan rooted at given `plan`.
+   * It requires that the given plan already gone through [[OptimizeSubqueries]] and the
+   * root node denoting a subquery is removed and optimized appropriately.
+   */
+  def transformWithSubqueries(plan: LogicalPlan)
```
[the remark this answers is not in the document]
Comment (vkorukanti) [reply core/src/main/scala/org/apache/spark/sql/delta/SubqueryTransformerHelper.scala:37]: updated the docs.

File: core/src/main/scala/org/apache/spark/sql/delta/stats/PrepareDeltaScan.scala (hunk 2 above)
Comment (zsxwing) [reply core/src/main/scala/org/apache/spark/sql/delta/stats/PrepareDeltaScan.scala:218]: NVM. I didn't notice we returned directly when hitting `V2WriteCommand`.

File: core/src/main/scala/org/apache/spark/sql/delta/PreprocessTableWithDVs.scala (hunk 1 above)
Comment (zsxwing) [reply core/src/main/scala/org/apache/spark/sql/delta/PreprocessTableWithDVs.scala:101]: Then this is fine.

**Review by zsxwing — APPROVED** [review]
LGTM

**Pull request closed without merging**